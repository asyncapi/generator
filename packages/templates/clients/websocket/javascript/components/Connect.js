import { Text } from '@asyncapi/generator-react-sdk';

export function Connect({ title }) {
  return (
    <Text newLines={2} indent={2}>
      {
        `// Method to establish a WebSocket connection
connect() {
  return new Promise((resolve, reject) => {
    this.websocket = new WebSocket(this.url);

    // On successful connection
    this.websocket.onopen = () => {
      console.log('Connected to ${title} server');
      resolve();
    };

    // On receiving a message
    this.websocket.onmessage = (event) => {
      if (this.messageHandlers.length > 0) {
        // Call custom message handlers
        this.messageHandlers.forEach(handler => {
          if (typeof handler === 'function') {
            this.handleMessage(event.data, handler);
          }
        });
      } else {
        // Default message logging
        console.log('Message received:', event.data);
      }
    };

    // On error first call custom error handlers, then default error behavior
    this.websocket.onerror = (error) => {
      if (this.errorHandlers.length > 0) {
        // Call custom error handlers
        this.errorHandlers.forEach(handler => handler(error));
      } else {
        // Default error behavior
        console.error('WebSocket Error:', error);
      }
      reject(error);
    };

    // On connection close
    this.websocket.onclose = () => {
      console.log('Disconnected from ${title} server');
    };
  });
}`
      }
    </Text>
  );
}
