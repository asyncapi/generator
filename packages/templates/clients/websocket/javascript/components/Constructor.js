import { Text } from '@asyncapi/generator-react-sdk';

export function Constructor({ serverUrl }) {
  return (
    <Text indent={2}>
      {
        `/*
  * Constructor to initialize the WebSocket client
  * @param {string} url - The WebSocket server URL. Use it if the server URL is different from the default one taken from the AsyncAPI document.
*/
constructor(url) {
  this.url = url || '${serverUrl}';
  this.websocket = null;
  this.messageHandlers = [];
  this.errorHandlers = [];
}
`
      }
    </Text>
  );
}
