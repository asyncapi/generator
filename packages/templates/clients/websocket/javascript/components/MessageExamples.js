import { getMessageExamples, getOperationMessages } from '@asyncapi/generator-helpers';
import { Text } from '@asyncapi/generator-react-sdk';

export default function MessageExamples({operation}) {
  const operationId = operation.id();
  const messageExamples = [];
  const messages = getOperationMessages(operation) || [];

  messages.forEach((message) => {
    const examples = getMessageExamples(message) || [];
    examples.forEach((example) => {
      const payload = example.payload();
      messageExamples.push(`\n\n**Example:**\n\`\`\`javascript\nclient.${operationId}(${JSON.stringify(payload, null, 2)});\n\`\`\``);
    });
  });
  return (
    <Text>
      {messageExamples.map(example => (
        <Text>
          {example}
        </Text>
      ))}
    </Text>
  );
}