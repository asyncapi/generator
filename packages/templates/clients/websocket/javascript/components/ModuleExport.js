import { Text } from '@asyncapi/generator-react-sdk';

export function ModuleExport({ clientName }) {
  return (
    <Text>
      {
        `module.exports = ${clientName};`
      }
    </Text>
  );
}
