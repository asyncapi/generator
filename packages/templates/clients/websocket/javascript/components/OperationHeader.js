import { Text } from '@asyncapi/generator-react-sdk';

export default function OperationHeader({operation}) {
  const operationId = operation.id();
  return (
    <Text>
      {`#### \`${operationId}(payload)\`
${operation.hasSummary() ?  operation.summary() : ''}
${operation.hasDescription() ? `\n${operation.description()}` : ''}`}
    </Text>
  );
}