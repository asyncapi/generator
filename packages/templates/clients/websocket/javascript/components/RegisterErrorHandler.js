import { Text } from '@asyncapi/generator-react-sdk';

export function RegisterErrorHandler() {
  return (
    <Text newLines={2} indent={2}>
      {
        `// Method to register custom error handlers
registerErrorHandler(handler) {
  if (typeof handler === 'function') {
    this.errorHandlers.push(handler);
  } else {
    console.warn('Error handler must be a function');
  }
}`
      }
    </Text>
  );
}
