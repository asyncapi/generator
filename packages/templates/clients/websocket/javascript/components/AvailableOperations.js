import { Text } from '@asyncapi/generator-react-sdk';
import OperationHeader from './OperationHeader';
import MessageExamples from './MessageExamples';

export function AvailableOperations({ operations }) {
  return (
    <>
      <Text newLines={2}>### Available Operations</Text>
      {operations.map((operation) => (
        <Text newLines={2}>
          <OperationHeader operation={operation} />
          <MessageExamples operation={operation} />
        </Text>
      ))}
    </>
  );
}
