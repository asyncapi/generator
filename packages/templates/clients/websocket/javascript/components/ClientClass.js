import { Text } from '@asyncapi/generator-react-sdk';
import { Constructor } from './Constructor';
import { Connect } from './Connect';
import { RegisterMessageHandler } from './RegisterMessageHandler';
import { RegisterErrorHandler } from './RegisterErrorHandler';
import { HandleMessage } from './HandleMessage';
import { SendEchoMessage } from './SendEchoMessage';
import { CloseConnection } from './CloseConnection';
import { ModuleExport } from './ModuleExport';

export function ClientClass({ clientName, serverUrl, title }) {
  return (
    <Text>
      <Text newLines={2}>
        {`class ${clientName} {`}
      </Text>
      <Constructor serverUrl={serverUrl} />
      <Connect title={title} />
      <RegisterMessageHandler />
      <RegisterErrorHandler />
      <HandleMessage />
      <SendEchoMessage />
      <CloseConnection />
      <Text>
        {'}'}
      </Text>
      <ModuleExport clientName={clientName} />
    </Text>
  );
}
