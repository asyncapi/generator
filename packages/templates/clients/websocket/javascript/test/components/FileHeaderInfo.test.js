import path from 'path';
import { Parser, fromFile } from '@asyncapi/parser';
import { FileHeaderInfo } from '../../components/FileHeaderInfo';

const parser = new Parser();
const asyncapi_websocket_query = path.resolve(__dirname, '../../../../../../helpers/test/__fixtures__/asyncapi-websocket-query.yml');

describe('Testing of FileHeaderInfo function', () => {
  let parsedAsyncAPIDocument;

  beforeAll(async () => {
    const parseResult = await fromFile(parser, asyncapi_websocket_query).parse();
    parsedAsyncAPIDocument = parseResult.document;
  });

  test('render websockets file header info correctly', () => {
    const wrapperWithPathname = FileHeaderInfo({ info: parsedAsyncAPIDocument.info(), server: parsedAsyncAPIDocument.servers().get('withPathname') });

    expect(wrapperWithPathname).toMatchSnapshot();

    const wrapperWithoutPathname = FileHeaderInfo({ info: parsedAsyncAPIDocument.info(), server: parsedAsyncAPIDocument.servers().get('withoutPathName') });

    expect(wrapperWithoutPathname).toMatchSnapshot();
  });
});
