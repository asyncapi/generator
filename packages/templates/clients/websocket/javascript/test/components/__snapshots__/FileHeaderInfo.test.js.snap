// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing of FileHeaderInfo function render websockets file header info with pathname correctly 1`] = `
"//////////////////////////////////////////////////////////////////////
//
// Gemini Market Data Websocket API - 1.0.0
// Protocol: wss
// Host: api.gemini.com
// Path: /v1/marketdata
//
//////////////////////////////////////////////////////////////////////"
`;

exports[`Testing of FileHeaderInfo function render websockets file header info without pathname correctly 1`] = `
"//////////////////////////////////////////////////////////////////////
//
// Gemini Market Data Websocket API - 1.0.0
// Protocol: wss
// Host: api.gemini.com
//
//////////////////////////////////////////////////////////////////////"
`;
